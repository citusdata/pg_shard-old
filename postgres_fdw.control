# postgres_fdw extension
comment = 'foreign-data wrapper for remote PostgreSQL servers'
default_version = 'develop'
module_pathname = '$libdir/postgres_fdw'
relocatable = true
