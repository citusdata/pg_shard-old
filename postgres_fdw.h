/*-------------------------------------------------------------------------
 *
 * postgres_fdw.h
 *		  Foreign-data wrapper for remote PostgreSQL servers
 *
 * Portions Copyright (c) 2012-2013, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		  contrib/postgres_fdw/postgres_fdw.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef POSTGRES_FDW_H
#define POSTGRES_FDW_H

#include "foreign/foreign.h"
#include "lib/stringinfo.h"
#include "nodes/relation.h"
#include "utils/rel.h"

#include "libpq-fe.h"

/* in postgres_fdw.c */
extern int	set_transmission_modes(void);
extern void reset_transmission_modes(int nestlevel);

/* in connection.c */
extern PGconn *GetConnection(ForeignServer *server, UserMapping *user,
			  bool will_prep_stmt);
extern void ReleaseConnection(PGconn *conn);
extern unsigned int GetCursorNumber(PGconn *conn);
extern unsigned int GetPrepStmtNumber(PGconn *conn);
extern void pgfdw_report_error(int elevel, PGresult *res, bool clear,
				   const char *sql);

/* in option.c */
extern int ExtractConnectionOptions(List *defelems,
						 const char **keywords,
						 const char **values);

/* in deparse.c */
extern void classifyConditions(PlannerInfo *root,
				   RelOptInfo *baserel,
				   List **remote_conds,
				   List **param_conds,
				   List **local_conds,
				   List **param_numbers);
extern void deparseSelectSql(StringInfo buf,
				 PlannerInfo *root,
				 RelOptInfo *baserel,
				 Bitmapset *attrs_used);
extern void appendWhereClause(StringInfo buf,
				  PlannerInfo *root,
				  List *exprs,
				  bool is_first);
extern void deparseInsertSql(StringInfo buf, PlannerInfo *root,
				 Index rtindex, Relation rel,
				 List *targetAttrs, List *returningList);
extern void deparseUpdateSql(StringInfo buf, PlannerInfo *root,
				 Index rtindex, Relation rel,
				 List *targetAttrs, List *returningList);
extern void deparseDeleteSql(StringInfo buf, PlannerInfo *root,
				 Index rtindex, Relation rel,
				 List *returningList);
extern void deparseAnalyzeSizeSql(StringInfo buf, Relation rel);
extern void deparseAnalyzeSql(StringInfo buf, Relation rel);

#endif   /* POSTGRES_FDW_H */
