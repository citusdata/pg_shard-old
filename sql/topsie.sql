--
--  Test topsie extension
--

CREATE EXTENSION topsie; -- fail, must install postgres_fdw first
CREATE EXTENSION postgres_fdw;

CREATE EXTENSION topsie SCHEMA public; -- fail, topsie is not relocatable

CREATE EXTENSION topsie;

-- Create shards for loopback server
CREATE TABLE widgets_s0 (
	c1 int PRIMARY KEY,
	c2 text
);
INSERT INTO topsie.dist_shards (shard_id, table_name) VALUES (0, 'widgets');

CREATE TABLE widgets_s1 (
	c1 int PRIMARY KEY,
	c2 text
);
INSERT INTO topsie.dist_shards (shard_id, table_name) VALUES (1, 'widgets');

-- Create main table on master
CREATE TABLE widgets (
	c1 int NOT NULL,
	c2 text
);

-- "Place" shards on loopback servers
INSERT INTO topsie.dist_shard_placements VALUES ('', (SELECT setting::integer FROM pg_settings WHERE name='port'), 1);
INSERT INTO topsie.dist_shard_placements VALUES ('', (SELECT setting::integer FROM pg_settings WHERE name='port'), 2);

-- Create servers
SELECT topsie.create_servers();

-- Enhance widgets table
SELECT topsie.rti_enhance_table('widgets', ARRAY['c1']);

-- Verify that selects query all shards
EXPLAIN (COSTS false) SELECT COUNT(*) FROM widgets;

-- Insert some data
INSERT INTO widgets VALUES ( 1, 'first row');
INSERT INTO widgets VALUES ( 2, 'second row');
INSERT INTO widgets VALUES ( 3, 'third row');
INSERT INTO widgets VALUES ( 4, 'fourth row');
INSERT INTO widgets VALUES ( 5, 'fifth row');
INSERT INTO widgets VALUES ( 6, 'sixth row');
INSERT INTO widgets VALUES ( 7, 'seventh row');
INSERT INTO widgets VALUES ( 8, 'eighth row');
INSERT INTO widgets VALUES ( 9, 'ninth row');
INSERT INTO widgets VALUES (10, 'last row');

-- Check that no duplicate data exists when querying all shards...
SELECT COUNT(DISTINCT c1) FROM widgets;

-- ... and that each shard has some data itself
SELECT COUNT(DISTINCT c1) FROM topsie.widgets_s0_r0;
SELECT COUNT(DISTINCT c1) FROM topsie.widgets_s1_r0;

-- Check that our table is now a view with a trigger
\dS widgets

-- Check for the foreign servers and tables.
SELECT COUNT(*) FROM pg_foreign_server;
SELECT COUNT(*) FROM pg_foreign_table;

DROP EXTENSION topsie; -- fail, because the trigger function lives in extension

DROP EXTENSION topsie CASCADE;
