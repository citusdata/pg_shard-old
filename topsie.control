# topsie extension
comment = 'foreign-data wrapper for sharding across remote PostgreSQL servers'
default_version = 'develop'
module_pathname = '$libdir/topsie'
relocatable = true
