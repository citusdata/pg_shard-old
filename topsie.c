#include "postgres.h"

#include "postgres_fdw.h"

#include "fmgr.h"
#include "access/htup_details.h"
#include "executor/executor.h"
#include "utils/builtins.h"
#include "utils/typcache.h"

static uint32 HashKeyForTuple(Datum hashKey, int16 attNum, TupleDesc tupDesc);

/*
 * topsie_hash accepts a record and attribute number and returns
 * the integer produced by hashing that attribute's value using the hash
 * function associated with the attribute's type. If no hash function can be
 * found for the attribute's type, an error occurs.
 */
PG_FUNCTION_INFO_V1(topsie_hash);

Datum
topsie_hash(PG_FUNCTION_ARGS) {
  HeapTupleHeader tup = PG_GETARG_HEAPTUPLEHEADER(0);
  int16 attNum = PG_GETARG_INT16(1);

  HeapTupleData tupleData = { 0 };
  Datum hashKey = 0;
  bool isNull = false;

  Oid typeId = HeapTupleHeaderGetTypeId(tup);
  int32 typMod = HeapTupleHeaderGetTypMod(tup);
  TupleDesc tupDesc = lookup_rowtype_tupdesc(typeId, typMod);

  /* Set struct fields: heap_getattr needs more than a bare HeapTupleHeader */
  tupleData.t_len = HeapTupleHeaderGetDatumLength(tup);
  ItemPointerSetInvalid(&(tupleData.t_self));
  tupleData.t_tableOid = InvalidOid;
  tupleData.t_data = tup;

  hashKey = heap_getattr(&tupleData, attNum, tupDesc, &isNull);

  /* Null values hash to zero */
  uint32 hashValue = isNull ? 0 : HashKeyForTuple(hashKey, attNum, tupDesc);

  ReleaseTupleDesc(tupDesc);

  PG_RETURN_UINT32(hashValue);
}

FunctionCallInfo
TopsieHashFnCallInfo(Oid typeId)
{
	TypeCacheEntry *typeEntry = NULL;
	FunctionCallInfo fcinfo = (FunctionCallInfo) palloc0(
			sizeof(FunctionCallInfoData));

	typeEntry = lookup_type_cache(typeId, TYPECACHE_HASH_PROC_FINFO);
	if (!OidIsValid(typeEntry->hash_proc_finfo.fn_oid))
	{
		ereport(ERROR,
				(errcode(ERRCODE_UNDEFINED_FUNCTION),
				 errmsg("could not identify a hash function for type %s",
				 format_type_be(typeId))));
	}

	InitFunctionCallInfoData(*fcinfo, &typeEntry->hash_proc_finfo, 1,
			InvalidOid, NULL, NULL);
	fcinfo->arg[0] = 0;
	fcinfo->argnull[0] = false;
	fcinfo->isnull = false;

	return fcinfo;
}

/*
 * HashKeyForTuple determines the type of the specified attribute within the
 * given tuple and uses that type information to look up an associated hash
 * function. The provided Datum is passed to that hash function to arrive at
 * a final hash value.
 */
static uint32
HashKeyForTuple(Datum hashKey, int16 attNum, TupleDesc tupDesc) {
  Oid attType = InvalidOid;
  FunctionCallInfo fcinfo = NULL;

  attType = tupDesc->attrs[attNum - 1]->atttypid;
  fcinfo = TopsieHashFnCallInfo(attType);

  fcinfo->arg[0] = hashKey;

  return DatumGetUInt32(FunctionCallInvoke(fcinfo));
}
