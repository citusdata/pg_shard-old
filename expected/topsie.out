--
--  Test topsie extension
--
CREATE EXTENSION topsie; -- fail, must install postgres_fdw first
ERROR:  required extension "postgres_fdw" is not installed
CREATE EXTENSION postgres_fdw;
CREATE EXTENSION topsie SCHEMA public; -- fail, topsie is not relocatable
ERROR:  extension "topsie" must be installed in schema "topsie"
CREATE EXTENSION topsie;
-- Create shards for loopback server
CREATE TABLE widgets_s0 (
	c1 int PRIMARY KEY,
	c2 text
);
INSERT INTO topsie.dist_shards (shard_id, table_name) VALUES (0, 'widgets');
CREATE TABLE widgets_s1 (
	c1 int PRIMARY KEY,
	c2 text
);
INSERT INTO topsie.dist_shards (shard_id, table_name) VALUES (1, 'widgets');
-- Create main table on master
CREATE TABLE widgets (
	c1 int NOT NULL,
	c2 text
);
-- "Place" shards on loopback servers
INSERT INTO topsie.dist_shard_placements VALUES ('', (SELECT setting::integer FROM pg_settings WHERE name='port'), 1);
INSERT INTO topsie.dist_shard_placements VALUES ('', (SELECT setting::integer FROM pg_settings WHERE name='port'), 2);
-- Create servers
SELECT topsie.create_servers();
 create_servers 
----------------
 
(1 row)

-- Enhance widgets table
SELECT topsie.rti_enhance_table('widgets', ARRAY['c1']);
 rti_enhance_table 
-------------------
 
(1 row)

-- Verify that selects query all shards
EXPLAIN (COSTS false) SELECT COUNT(*) FROM widgets;
                QUERY PLAN                 
-------------------------------------------
 Aggregate
   ->  Append
         ->  Foreign Scan on widgets_s0_r0
         ->  Foreign Scan on widgets_s1_r0
(4 rows)

-- Insert some data
INSERT INTO widgets VALUES ( 1, 'first row');
INSERT INTO widgets VALUES ( 2, 'second row');
INSERT INTO widgets VALUES ( 3, 'third row');
INSERT INTO widgets VALUES ( 4, 'fourth row');
INSERT INTO widgets VALUES ( 5, 'fifth row');
INSERT INTO widgets VALUES ( 6, 'sixth row');
INSERT INTO widgets VALUES ( 7, 'seventh row');
INSERT INTO widgets VALUES ( 8, 'eighth row');
INSERT INTO widgets VALUES ( 9, 'ninth row');
INSERT INTO widgets VALUES (10, 'last row');
-- Check that no duplicate data exists when querying all shards...
SELECT COUNT(DISTINCT c1) FROM widgets;
 count 
-------
    10
(1 row)

-- ... and that each shard has some data itself
SELECT COUNT(DISTINCT c1) FROM topsie.widgets_s0_r0;
 count 
-------
     5
(1 row)

SELECT COUNT(DISTINCT c1) FROM topsie.widgets_s1_r0;
 count 
-------
     5
(1 row)

-- Check that our table is now a view with a trigger
\dS widgets
    View "public.widgets"
 Column |  Type   | Modifiers 
--------+---------+-----------
 c1     | integer | not null
 c2     | text    | 
Triggers:
    dist_insert INSTEAD OF INSERT ON widgets FOR EACH ROW EXECUTE PROCEDURE topsie.t001_shard_insert('2', '1', 'c1')

-- Check for the foreign servers and tables.
\des
                            List of foreign servers
                      Name                      | Owner | Foreign-data wrapper 
------------------------------------------------+-------+----------------------
 topsie_worker-98f309c260d2c990f9b1a9afe75c486c | jason | postgres_fdw
(1 row)

\det
                         List of foreign tables
 Schema |     Table     |                     Server                     
--------+---------------+------------------------------------------------
 topsie | widgets_s0_r0 | topsie_worker-98f309c260d2c990f9b1a9afe75c486c
 topsie | widgets_s1_r0 | topsie_worker-98f309c260d2c990f9b1a9afe75c486c
(2 rows)

DROP EXTENSION topsie; -- fail, because the trigger function lives in extension
ERROR:  cannot drop extension topsie because other objects depend on it
DETAIL:  trigger dist_insert on view widgets depends on function topsie.t001_shard_insert()
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP EXTENSION topsie CASCADE;
NOTICE:  drop cascades to trigger dist_insert on view widgets
